### PR TITLE
Support instance masks (N,H,W)

### LINKED ISSUE #2855: `AugmentationSequential` does not support instance masks shape (N, H, W)
### Describe the bug

Currently `AugmentationSequential` does not seem to support instance masks i.e. an image (C, H, W) having multiple masks (N, H, W). 

### Reproduction steps

The following code snippet errors out with a `RuntimeError: Expect (B, C, H, W). Got torch.Size([3, 256, 256]).` error.

```python
import torch
import kornia.augmentation as K

img = torch.randn(3, 256, 256)
masks = torch.ones( 2, 256, 256)

aug_dict = K.AugmentationSequential(
    K.RandomHorizontalFlip(p=1.0),
    data_keys=None,
    keepdim=True,
)
aug_dict.keepdim = True

data = {'image': img, 'mask': masks}
out = aug_dict(data)
```

### Expected behavior

Error should not be raised

### Environment

```shell
wget https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py

kornia==0.7.2
```
- PyTorch Version (e.g., 1.0):
- OS (e.g., Linux):
- How you installed PyTorch (`conda`, `pip`, source):
- Build command you used (if compiling from source):
- Python version:
- CUDA/cuDNN version:
- GPU models and configuration:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by ashnair1**
#### Changes

Fixes #2855 related to #941

#### Type of change

- [x] 🧪 Tests Cases
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: tests/augmentation/test_container.py [hunk 1]
```diff
@@ -524,7 +524,7 @@ def test_class(self, device, dtype):
 
     @pytest.mark.parametrize("B", [None, 1, 3, 5])
     @pytest.mark.parametrize("C_i", [1, 3])
-    @pytest.mark.parametrize("C_m", [None, 1])
+    @pytest.mark.parametrize("C_m", [None, 1, 3])
```
Comment (johnnv1) [inline tests/augmentation/test_container.py:527]:
```suggestion
    @pytest.mark.parametrize("C_m", [None, 1, 3, 4])
```

File: kornia/augmentation/container/ops.py [hunk 2]
```diff
@@ -286,10 +287,11 @@ def transform_list(
         if isinstance(module, (K.GeometricAugmentationBase2D,)):
             tfm_input = []
             params = cls.get_instance_module_param(param)
-            for inp in input:
-                inp = module.transform_tensor(inp)
+            params_i = copy.deepcopy(params)
```
Comment (edgarriba) [inline kornia/augmentation/container/ops.py:290]: Why do we need to make a copy here ?

**Review by edgarriba — APPROVED** [review]

File: kornia/augmentation/container/ops.py (hunk 2 above)
Comment (ashnair1) [reply kornia/augmentation/container/ops.py:290]: To iteratively modify the batch_prob.

In batched cases, image would a tensor of shape (B, C, H, W) whereas masks would be a list of tensors [(N1, H, W), (N2, H, W)....].

`params["batch_prob"]` is set w.r.t to the image tensor, so we need to iterate over it and reassign the batch_prob passed while keeping the rest of the params the same

Comment (shijianjian) [reply kornia/augmentation/container/ops.py:290]: I think there should at least have an `if-else` block here. If it is a list, we enumerate them one-by-one. If it is a straight-away tensor, we do it all at once.

**Review by shijianjian — CHANGES_REQUESTED** [review]
Meanwhile, what we did for the boxes is to merge them all into one tensor, which is
https://github.com/kornia/kornia/blob/46a5e40a0dbffcb38c7eca5b36b39479dcc0d85f/kornia/geometry/boxes.py#L20-L33

Not sure if we should do the same to the masks. Can anyone benchmark on which approach runs faster? Iteration or batching?

File: kornia/augmentation/container/ops.py (hunk 2 above)
Comment (ashnair1) [reply kornia/augmentation/container/ops.py:290]: Not sure I follow. This change takes place in the `transform_list` function of `MaskSequentialOps`. So the input is a list of mask tensors.

**Comment by ashnair1** [comment]
Merging a list of tensors into tensors will not work for instance segmentation. This is because each image will have a different number of objects and so different number of masks. Because of this, they cannot be batched. 

This is how it would look like:
**Single sample**: Image (C, H, W) and Mask (N, H, W)
**Batched**: Image (B, C, H, W) and Mask ([(N<sub>1</sub>, H, W), (N<sub>2</sub>, H, W) ... (N<sub>b</sub>, H, W)]) 

`N` (no. of detections) varies from image to image, so they cannot be batched.

File: kornia/augmentation/container/ops.py (hunk 2 above)
Comment (shijianjian) [reply kornia/augmentation/container/ops.py:290]: Ah, this should be fine then.

**Comment by shijianjian** [comment]
> Single sample: Image (C, H, W) and Mask (N, H, W)
> Batched: Image (B, C, H, W) and Mask ([(N1, H, W), (N2, H, W) ... (Nb, H, W)])

I meant that
```python
N_max = max(N1, ..., Nb)
padding = [N_max - N1, ..., N_max - Nb]
padded_masks = (B, N_max, H, W)

... # After augmentation

Unpad (B, N_max, H, W) => Mask ([(N1, H, W), (N2, H, W) ... (Nb, H, W)])
```

The current change is fine in this PR I think. We should benchmark the iterative and batching strategies to see how those methods perform. Maybe a `Mask` class, similar to `Boxes`, to handle these.

**Review by shijianjian — APPROVED** [review]

**Review by johnnv1 — APPROVED** [review]

**Pull request merged**